codewof/tests/codewof_test_data_generator.py

```diff
@@ -86,6 +88,233 @@ def generate_users(user):
+        description='Few know of this group...'
+    )
+    group_6 = Group.objects.create(
+        name='Team 300',
+        description='Rite of Passage'
+    )
+    group_7 = Group.objects.create(
+        name='Team CSERG',
+        description='We made CodeWOF.'
+    )
+    group_8 = Group.objects.create(
+        name='Class 1',
+        description='The Group for Class 1.'
+    )
+
+    group_1.save()
+    group_2.save()
+    group_3.save()
+    group_4.save()
+    group_5.save()
+    group_6.save()
+    group_7.save()
+    group_8.save()
+
+
+def generate_memberships():
+    """Generate memberships for codeWOF tests. Memberships are generated for users 1, 2, and 3, and every group created
+    in generate_groups, covering all the GroupRoles."""
+    group_north = Group.objects.get(name='Group North')
+    group_east = Group.objects.get(name='Group East')
+    group_west = Group.objects.get(name='Group West')
+    group_south = Group.objects.get(name='Group South')
+    management.call_command("load_group_roles")
+    user1 = User.objects.get(id=1)
+    user2 = User.objects.get(id=2)
+    user3 = User.objects.get(id=3)
+    admin_role = GroupRole.objects.get(name='Admin')
+    member_role = GroupRole.objects.get(name='Member')
+
+    membership_1 = Membership.objects.create(
+        user=user1,
+        group=group_north,
+        role=admin_role
+    )
+    membership_2 = Membership.objects.create(
+        user=user1,
+        group=group_east,
+        role=member_role
+    )
+    membership_3 = Membership.objects.create(
+        user=user1,
+        group=group_west,
+        role=member_role
+    )
+    membership_4 = Membership.objects.create(
+        user=user1,
+        group=group_south,
+        role=member_role
+    )
+    membership_5 = Membership.objects.create(
+        user=user2,
+        group=group_north,
+        role=member_role
+    )
+    membership_6 = Membership.objects.create(
+        user=user3,
+        group=group_north,
+        role=member_role
+    )
+
+    membership_1.save()
+    membership_2.save()
+    membership_3.save()
+    membership_4.save()
+    membership_5.save()
+    membership_6.save()
+
+
+def generate_email_accounts():
+    """Generates email accounts for user 1 and 2 for codeWOF tests."""
+    user1 = User.objects.get(id=1)
+    user2 = User.objects.get(id=2)
+    email1 = EmailAddress(
+        user=user1,
+        email=user1.email,
+        primary=True,
+        verified=True
+    )
+    email2 = EmailAddress(
+        user=user1,
+        email="john@mail.com",
+        primary=False,
+        verified=True
+    )
+    email3 = EmailAddress(
+        user=user1,
+        email="jack@mail.com",
+        primary=False,
+        verified=False
+    )
+    email4 = EmailAddress(
+        user=user2,
+        email=user2.email,
+        primary=True,
+        verified=True
+    )
+    email1.save()
+    email2.save()
+    email3.save()
+    email4.save()
+
+
+def generate_invitations():
```

Comment:
It appears this file setups up the database with a whole bunch of preset data for testing. It is fine for this project, but I would recommend future test suites have a test data generator that allows different data to be set for each test.

See the following links for how we use these in CS Unplugged:
- https://github.com/uccser/cs-unplugged/blob/develop/csunplugged/tests/topics/TopicsTestDataGenerator.py
- https://github.com/uccser/cs-unplugged/blob/develop/csunplugged/tests/topics/views/test_lesson_view.py#L55-L72

Reply:
Okay, I'll leave it for the time being.